docs/src/adr/015adr-trace.md

```diff
@@ -0,0 +1,403 @@
+ strict rule about when to use sequences or tuples. Apalache differentiates
+ between tuples and sequences, and it may produce both forms of expressions.
+
+ 1. A set of the form `{ "#set": [ <expr>, ..., <expr> ] }`. A set is different
+ from a list in that it does not assume any ordering of its elements. However,
+ it is only a syntax form in our format. Apalache distinguishes between sets
+ and lists and thus it will output sets in the set form. Other tools may
+ interpret sets as lists.
+
+ 1. A map of the form `{ "#map": [ [ <expr>, <expr> ], ..., [ <expr>, <expr> ]
+ ] }`.  That is, a map holds a JSON array of two-element arrays. Each
+ two-element array `p` is interpreted as follows: `p[0]` is the map key and
+ `p[1]` is the map value. Importantly, a key may be an arbitrary expression. It
+ does not have to be a string or an integer. TLA+ functions are written as maps
+ in this format.
+
+
+### Example
+
+The counterexample to `NoSolution` may be written in the ITF format as follows:
+
+```js
+{
+  "#meta": {
+    "source": "MC_MissionariesAndCannibalsTyped.tla"
+  },
+  "vars": [ "bank_of_boat", "who_is_on_bank" ],
+  "states": [
+    {
+      "#meta": { "index": 0 },
+      "bank_of_boat": "E",
+      "who_is_on_bank": {
+        "#map": [
+          [ "E", { "#set": [ "c1_OF_PERSON", "c2_OF_PERSON",
+                             "m1_OF_PERSON", "m2_OF_PERSON" ] } ],
+          [ "W", { "#set": [] } ]
+        ]  
+      }
+    },
+    {
+      "#meta": { "index": 1 },
+      "bank_of_boat": "W",
+      "who_is_on_bank": {
+        "#map": [
+          [ "E", { "#set": [ "c1_OF_PERSON", "m1_OF_PERSON" ] } ],
+          [ "W", { "#set": [ "c2_OF_PERSON", "m2_OF_PERSON" ] } ]
+        ]  
+      }
+    },
+    {
+      "#meta": { "index": 2 },
+      "bank_of_boat": "E",
+      "who_is_on_bank": {
+        "#map": [
+          [ "E", { "#set": [ "c1_OF_PERSON",
+                             "m1_OF_PERSON", "m2_OF_PERSON" ] } ],
+          [ "W", { "#set": [ "c2_OF_PERSON" ] } ]
+        ]  
+      }
+    },
+    {
+      "#meta": { "index": 3 },
+      "bank_of_boat": "W",
+      "who_is_on_bank": {
+        "#map": [
+          [ "E", { "#set": [ "c1_OF_PERSON" ] } ],
+          [ "W", { "#set": [ "c2_OF_PERSON", "m1_OF_PERSON", "m2_OF_PERSON" ] } ]
+        ]  
+      }
+    },
+    {
+      "#meta": { "index": 4 },
+      "bank_of_boat": "E",
+      "who_is_on_bank": {
+        "#map": [
+          [ "E", { "#set": [ "c1_OF_PERSON", "c2_OF_PERSON" ] } ],
+          [ "W", { "#set": [ "m1_OF_PERSON", "m2_OF_PERSON" ] } ]
+        ]  
+      }
+    },
+    {
+      "#meta": { "index": 5 },
+      "bank_of_boat": "W",
+      "who_is_on_bank": {
+        "#map": [
+          [ "E", { "#set": [ ] } ],
+          [ "W", { "#set": [ "c1_OF_PERSON", "c2_OF_PERSON",
+                             "m1_OF_PERSON", "m2_OF_PERSON" ] } ]
+        ]  
+      }
+    }
+  ]
+}
+```
```

Comment:
In contrast to the other format, it looks like this one does not include data about the invariant that was violated. It'd think that was important info. Should it be included?

Reply:
A trace is a general concept, it doesn't have to come from a counterexample to an invariant.

Reply:
That makes sense. But the main motivating force for this to report on counter examples, and those are used as the examples. So it seems like we should also account for how we’ll supply this data when needed, and probably make a clearer statement about this being a general form for traces which can be specialized to meet the needs of reporting counter examples. (Especially since this linked from our public docs).

Reply:
Generally speaking, at least for the audit case, the invariant isn't referenced by any of the tools that consume counterexample traces. But since this is JSON, you're always allowed to add non-specified fields as you wish, it'd be easy enough to add a toplevel `"(violated) invariant" : expr` entry.

Reply:
This can go into `#meta`. I am thinking of this format as of a trace format, not being really specific to TLA+.

Reply:
Adding info on this one way or another would help, even if just to say: "we don't actually need to include the invariant".